File: go/exercises/practice/forth/forth.go
package forth

import (
	"errors"
	"strconv"
	"strings"
)

type Evaluator struct {
	stack []int
	words map[string][]string
}

func NewEvaluator() *Evaluator {
	return &Evaluator{
		stack: []int{},
		words: make(map[string][]string),
	}
}

func (e *Evaluator) evaluateTokens(tokens []string) error {
	i := 0
	for i < len(tokens) {
		token := strings.ToLower(tokens[i])
		if token == ":" {
			// Definition: : name body... ;
			if i+1 >= len(tokens) {
				return errors.New("empty definition")
			}
			name := strings.ToLower(tokens[i+1])
			// Check if name is a number
			if _, err := strconv.Atoi(name); err == nil {
				return errors.New("illegal operation")
			}
			i += 2
			bodyStart := i
			for i < len(tokens) && strings.ToLower(tokens[i]) != ";" {
				i++
			}
			if i >= len(tokens) {
				return errors.New("unterminated definition")
			}
			body := tokens[bodyStart:i]

			// Save old definition of name if any
			oldBody, hasOld := e.words[name]

			resolvedBody := e.expandBody(body, name, oldBody, hasOld)
			e.words[name] = resolvedBody

			i++ // skip ";"
		} else {
			if err := e.executeToken(token); err != nil {
				return err
			}
			i++
		}
	}
	return nil
}

func (e *Evaluator) expandBody(body []string, currentName string, oldBody []string, hasOld bool) []string {
	var result []string
	for _, t := range body {
		tLower := strings.ToLower(t)
		if tLower == currentName {
			if hasOld {
				result = append(result, e.expandBody(oldBody, "", nil, false)...)
			} else {
				result = append(result, tLower)
			}
		} else if sub, ok := e.words[tLower]; ok {
			result = append(result, e.expandBody(sub, "", nil, false)...)
		} else {
			result = append(result, tLower)
		}
	}
	return result
}

func (e *Evaluator) executeToken(token string) error {
	// 1. Is it a number?
	if val, err := strconv.Atoi(token); err == nil {
		e.stack = append(e.stack, val)
		return nil
	}

	// 2. Is it a user-defined word?
	if body, ok := e.words[token]; ok {
		for _, bt := range body {
			if err := e.executeToken(bt); err != nil {
				return err
			}
		}
		return nil
	}

	// 3. Built-in operations
	switch token {
	case "+":
		if len(e.stack) < 2 {
			if len(e.stack) == 0 {
				return errors.New("empty stack")
			}
			return errors.New("only one value on the stack")
		}
		b := e.stack[len(e.stack)-1]
		a := e.stack[len(e.stack)-2]
		e.stack = e.stack[:len(e.stack)-2]
		e.stack = append(e.stack, a+b)
	case "-":
		if len(e.stack) < 2 {
			if len(e.stack) == 0 {
				return errors.New("empty stack")
			}
			return errors.New("only one value on the stack")
		}
		b := e.stack[len(e.stack)-1]
		a := e.stack[len(e.stack)-2]
		e.stack = e.stack[:len(e.stack)-2]
		e.stack = append(e.stack, a-b)
	case "*":
		if len(e.stack) < 2 {
			if len(e.stack) == 0 {
				return errors.New("empty stack")
			}
			return errors.New("only one value on the stack")
		}
		b := e.stack[len(e.stack)-1]
		a := e.stack[len(e.stack)-2]
		e.stack = e.stack[:len(e.stack)-2]
		e.stack = append(e.stack, a*b)
	case "/":
		if len(e.stack) < 2 {
			if len(e.stack) == 0 {
				return errors.New("empty stack")
			}
			return errors.New("only one value on the stack")
		}
		b := e.stack[len(e.stack)-1]
		a := e.stack[len(e.stack)-2]
		if b == 0 {
			return errors.New("divide by zero")
		}
		e.stack = e.stack[:len(e.stack)-2]
		e.stack = append(e.stack, a/b)
	case "dup":
		if len(e.stack) < 1 {
			return errors.New("empty stack")
		}
		top := e.stack[len(e.stack)-1]
		e.stack = append(e.stack, top)
	case "drop":
		if len(e.stack) < 1 {
			return errors.New("empty stack")
		}
		e.stack = e.stack[:len(e.stack)-1]
	case "swap":
		if len(e.stack) < 2 {
			if len(e.stack) == 0 {
				return errors.New("empty stack")
			}
			return errors.New("only one value on the stack")
		}
		n := len(e.stack)
		e.stack[n-1], e.stack[n-2] = e.stack[n-2], e.stack[n-1]
	case "over":
		if len(e.stack) < 2 {
			if len(e.stack) == 0 {
				return errors.New("empty stack")
			}
			return errors.New("only one value on the stack")
		}
		second := e.stack[len(e.stack)-2]
		e.stack = append(e.stack, second)
	default:
		return errors.New("undefined operation")
	}

	return nil
}

func Forth(input []string) ([]int, error) {
	eval := NewEvaluator()
	for _, line := range input {
		fields := strings.Fields(line)
		if err := eval.evaluateTokens(fields); err != nil {
			return nil, err
		}
	}
	return eval.stack, nil
}

File: javascript/exercises/practice/forth/forth.js
//
// This is only a SKELETON file for the 'Forth' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Forth {
  constructor() {
    this._stack = [];
    this.dictionary = new Map();

    // Register built-in words
    this.dictionary.set('+', (stack) => {
      if (stack.length < 2) throw new Error('Stack empty');
      const b = stack.pop();
      const a = stack.pop();
      stack.push(a + b);
    });

    this.dictionary.set('-', (stack) => {
      if (stack.length < 2) throw new Error('Stack empty');
      const b = stack.pop();
      const a = stack.pop();
      stack.push(a - b);
    });

    this.dictionary.set('*', (stack) => {
      if (stack.length < 2) throw new Error('Stack empty');
      const b = stack.pop();
      const a = stack.pop();
      stack.push(a * b);
    });

    this.dictionary.set('/', (stack) => {
      if (stack.length < 2) throw new Error('Stack empty');
      const b = stack.pop();
      const a = stack.pop();
      if (b === 0) throw new Error('Division by zero');
      stack.push(Math.trunc(a / b));
    });

    this.dictionary.set('dup', (stack) => {
      if (stack.length < 1) throw new Error('Stack empty');
      stack.push(stack[stack.length - 1]);
    });

    this.dictionary.set('drop', (stack) => {
      if (stack.length < 1) throw new Error('Stack empty');
      stack.pop();
    });

    this.dictionary.set('swap', (stack) => {
      if (stack.length < 2) throw new Error('Stack empty');
      const b = stack.pop();
      const a = stack.pop();
      stack.push(b);
      stack.push(a);
    });

    this.dictionary.set('over', (stack) => {
      if (stack.length < 2) throw new Error('Stack empty');
      stack.push(stack[stack.length - 2]);
    });
  }

  evaluate(text) {
    const tokens = this._tokenize(text);
    let i = 0;
    while (i < tokens.length) {
      const token = tokens[i];
      if (token === ':') {
        i++;
        if (i >= tokens.length) throw new Error('Invalid definition');
        const wordName = tokens[i].toLowerCase();
        if (!isNaN(Number(wordName))) {
          throw new Error('Invalid definition');
        }
        i++;
        const body = [];
        let closed = false;
        while (i < tokens.length) {
          const t = tokens[i];
          if (t === ';') {
            closed = true;
            i++;
            break;
          } else {
            body.push(t);
            i++;
          }
        }
        if (!closed) throw new Error('Invalid definition');

        // Compile body: resolve existing dictionary words at definition time
        const compiledBody = this._compileBody(body);
        this.dictionary.set(wordName, compiledBody);
      } else {
        this._executeToken(token, this._stack);
        i++;
      }
    }
  }

  _tokenize(text) {
    return text.trim().split(/\s+/).filter(Boolean);
  }

  _compileBody(bodyTokens) {
    const result = [];
    for (const t of bodyTokens) {
      const lowerT = t.toLowerCase();
      if (!isNaN(Number(t))) {
        result.push(t);
      } else if (this.dictionary.has(lowerT)) {
        const entry = this.dictionary.get(lowerT);
        if (Array.isArray(entry)) {
          result.push(...entry);
        } else {
          result.push(lowerT);
        }
      } else {
        result.push(lowerT);
      }
    }
    return result;
  }

  _executeTokens(tokens, stack) {
    for (const t of tokens) {
      this._executeToken(t, stack);
    }
  }

  _executeToken(token, stack) {
    const lower = token.toLowerCase();
    if (!isNaN(Number(token))) {
      stack.push(Number(token));
    } else if (this.dictionary.has(lower)) {
      const entry = this.dictionary.get(lower);
      if (typeof entry === 'function') {
        entry(stack);
      } else if (Array.isArray(entry)) {
        this._executeTokens(entry, stack);
      }
    } else {
      throw new Error('Unknown command');
    }
  }

  get stack() {
    return [...this._stack];
  }
}
